File: programs/matmul/matmul.c
#include <stdio.h>
#include <stdlib.h>

int main(void) {
    const int n = 256;
    const int nn = n * n;
    int *a = malloc((size_t)nn * sizeof(int));
    int *b = malloc((size_t)nn * sizeof(int));
    int *c = malloc((size_t)nn * sizeof(int));
    for (int i = 0; i < nn; i++) {
        int r = i / n;
        int col = i - r * n;
        a[i] = ((r * 7 + col * 13) % 16) - 8;
        b[i] = ((r * 13 - col * 7) % 16) - 8;
        c[i] = 0;
    }
    for (int r = 0; r < n; r++) {
        for (int k = 0; k < n; k++) {
            int aik = a[r * n + k];
            for (int j = 0; j < n; j++) {
                c[r * n + j] += aik * b[k * n + j];
            }
        }
    }
    long sum = 0;
    for (int q = 0; q < nn; q++) sum += c[q];
    printf("matmul checksum = %ld\n", sum);
    free(a); free(b); free(c);
    return 0;
}

File: programs/matmul/matmul.dart
void main() {
  const n = 256;
  final nn = n * n;
  final a = List<int>.filled(nn, 0);
  final b = List<int>.filled(nn, 0);
  final c = List<int>.filled(nn, 0);
  var i = 0;
  while (i < nn) {
    final r = i ~/ n;
    final col = i - r * n;
    a[i] = ((r * 7 + col * 13).remainder(16)) - 8;
    b[i] = ((r * 13 - col * 7).remainder(16)) - 8;
    i++;
  }
  var r = 0;
  while (r < n) {
    var k = 0;
    while (k < n) {
      final aik = a[r * n + k];
      var j = 0;
      while (j < n) {
        c[r * n + j] += aik * b[k * n + j];
        j++;
      }
      k++;
    }
    r++;
  }
  var sum = 0;
  var q = 0;
  while (q < nn) {
    sum += c[q];
    q++;
  }
  print("matmul checksum = $sum");
}
